Mathlib/RingTheory/Presentation.lean

```diff
@@ -0,0 +1,224 @@
+/-
+Copyright (c) 2024 Christian Merten. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jung Tao Cheng, Christian Merten, Andrew Yang
+-/
+import Mathlib.LinearAlgebra.TensorProduct.RightExactness
+import Mathlib.RingTheory.FinitePresentation
+import Mathlib.RingTheory.Generators
+import Mathlib.RingTheory.MvPolynomial.Localization
+import Mathlib.RingTheory.TensorProduct.MvPolynomial
+
+/-!
+
+# Presentations of algebras
+
+A presentation of an `R`-algebra `S` is a distinguished family of generators and relations.
+
+## Main definition
+
+- `Algebra.Presentation`: A presentation of an `R`-algebra `S` is a family of
+  generators with
+  1. `rels`: The type of relations.
+  2. `relation : relations → MvPolynomial vars R`: The assignment of
+     each relation to a polynomial in the generators.
+- `Algebra.Presentation.IsFinite`: A presentation is called finite, if both variables and relations
+  are finite.
+- `Algebra.Presentation.dimension`: The dimension of a presentation is the number of generators
+  minus the number of relations.
+
+We also give constructors for localization and base change.
+
+## TODO
+
+- Define composition of presentations.
+- Define `Hom`s of presentations.
+
+-/
+
+universe t w u v
+
+open TensorProduct MvPolynomial
+
+variable (R : Type u) (S : Type v) [CommRing R] [CommRing S] [Algebra R S]
+
+/--
+A presentation of an `R`-algebra `S` is a family of
+generators with
+1. `rels`: The type of relations.
+2. `relation : relations → MvPolynomial vars R`: The assignment of
+each relation to a polynomial in the generators.
+-/
+@[nolint checkUnivs]
+structure Algebra.Presentation extends Algebra.Generators.{w} R S where
+  /-- The type of relations.  -/
+  rels : Type t
+  /-- The assignment of each relation to a polynomial in the generators. -/
+  relation : rels → MvPolynomial vars R
+  /-- The relations span the kernel of the canonical map. -/
+  ker_algebraMap_eq_span_range_relation :
+    RingHom.ker (aeval val) = Ideal.span (Set.range relation)
```

Comment:
```suggestion
    toGenerators.ker = Ideal.span (Set.range relation)
```
I wonder if we should just call it `span_range_relation_eq_ker` and replace the lemma below.

Reply:
You mean, state this the other way round?

Reply:
Yes.

Reply:
Done.